openapi3/schema_formats.go

```diff
@@ -21,26 +21,111 @@ type FormatCallback func(value string) error
+	if format != nil {
+		if format.versionedFormats == nil {
+			format.versionedFormats = make([]*versionedFormat, minMinorVersion+1)
+			format.versionedFormats[minMinorVersion] = vFormat
+		} else {
+			numVersionedFormats := uint64(len(format.versionedFormats))
+			if minMinorVersion >= numVersionedFormats {
+				// grow array
+				lastValue := format.versionedFormats[numVersionedFormats-1]
+				additionalEntries := make([]*versionedFormat, minMinorVersion+1-numVersionedFormats)
+				if lastValue != nil {
+					for i := 0; i < len(additionalEntries); i++ {
+						additionalEntries[i] = lastValue
+					}
+				}
+				format.versionedFormats = append(format.versionedFormats, additionalEntries...)
+				format.versionedFormats[minMinorVersion] = vFormat
+				return
+			}
+			if override {
+				for i := minMinorVersion; i < numVersionedFormats; i++ {
+					format.versionedFormats[i] = vFormat
+				}
+
+			} else {
+				minValue := format.versionedFormats[minMinorVersion]
+				for i := minMinorVersion; i < numVersionedFormats && minValue == format.versionedFormats[i]; i++ {
+					format.versionedFormats[i] = vFormat
+				}
+
+			}
+		}
+	}
+}
+
+func (format Format) get(minorVersion uint64) *versionedFormat {
+	if format.versionedFormats != nil {
+		if minorVersion >= uint64(len(format.versionedFormats)) {
+			return format.versionedFormats[len(format.versionedFormats)-1]
+		}
+		return format.versionedFormats[minorVersion]
+	}
+	return nil
+}
+
+func (format Format) DefinedForMinorVersion(minorVersion uint64) bool {
+	return format.get(minorVersion) != nil
+}
+
 // SchemaStringFormats allows for validating string formats
-var SchemaStringFormats = make(map[string]Format, 4)
+var SchemaStringFormats = make(map[string]*Format, 4)
 
 // DefineStringFormat defines a new regexp pattern for a given format
+// Will enforce regexp usage for minor versions of OpenAPI (3.Y.Z)
 func DefineStringFormat(name string, pattern string) {
+	DefineStringFormatStartingWithOpenAPIMinorVersion(name, 0, pattern, true)
+}
+
+func getSchemaStringFormats(name string, minorVersion uint64) *versionedFormat {
+	if currentStringFormat, found := SchemaStringFormats[name]; found {
+		return currentStringFormat.get(minorVersion)
+	}
+	return nil
+}
+
+func updateSchemaStringFormats(name string, minMinorVersion uint64, vFormat *versionedFormat, override bool) {
+	if currentStringFormat, found := SchemaStringFormats[name]; found {
+		currentStringFormat.add(minMinorVersion, vFormat, override)
+		return
+	}
+	var newFormat Format
+	newFormat.add(minMinorVersion, vFormat, override)
+	SchemaStringFormats[name] = &newFormat
+}
+
+// DefineStringFormatStartingWithOpenAPIMinorVersion defines a new regexp pattern for a given format to apply for a minor version of OpenAPI 3
+// override allows to enforce regex usage to all subsequent minor versions of OpenAPI
+func DefineStringFormatStartingWithOpenAPIMinorVersion(name string, fromMinorVersion uint64, pattern string, override bool) {
```

Comment:
Why `uint64`? `uint8` or even `uint16` is large enough. 

I like the idea but I'd rather have a simpler API:
* a shorter method such as the existing `DefineStringFormat` (extend with variadic args)
* make `override=true` the only behavior. API users will simply make multiple calls to cut specific ranges
* turn `fromMinorVersion` into a `float` (`f32` if for all `Y`, `3.Y` can be expressed and I believe it's a yes) so users can see `DefineStringFormat(name, pattern, 3.1)`. The call should panic on non-nil non-floats and compare minors with a `0.4` epsilon.

Reply:
(yes, I'm asking to extend the existing 2 methods with a variadic float)

Reply:
uint64 comes from the string parsing function used when parsing OpenAPI document 
=> Go string parsing returns a uint64
=> Casting to uint8 would require minor version

Since most processors are 64 bits, using an uint8/uint64 does save memory nor it brings performance gain.

Reply:
Gone with the variadic options. have a look